File: clojure/allergies/src/allergies.clj
(ns allergies)

(def allergen->score
  {:eggs 0
   :peanuts 1
   :shellfish 2
   :strawberries 3
   :tomatoes 4
   :chocolate 5
   :pollen 6
   :cats 7})

(defn allergies [score]
  (keep
   (fn [[allergen i]]
     (when (bit-test score i) allergen))
   allergen->score))

(defn allergic-to? [score allergen]
  (bit-test score (allergen allergen->score)))

File: ruby/allergies/allergies.rb
class Allergies
  ALLERGEN_BIT = {
    0 => 'eggs',
    1 => 'peanuts',
    2 => 'shellfish',
    3 => 'strawberries',
    4 => 'tomatoes',
    5 => 'chocolate',
    6 => 'pollen',
    7 => 'cats'
  }

  def allergic_to?(food)
    allergies.include?(food)
  end

  def list
    allergies
  end

  private

  attr_reader :allergies

  def initialize(type)
    allergies = []
    type.bit_length.downto(0) do |i|
      allergies << ALLERGEN_BIT[i] if type[i] == 1
    end
    @allergies = allergies.reject(&:nil?)
  end
end
